File: ois/ois20/round2/ois11_round2_solutions/ransomware.cpp
// Starting from position 0, determines the (unique) match and advances
// the current position by the length of the matched codeword.

#include <iostream>
#include <cstring>
#include <cassert>

using namespace std;
const int MAXN = 100;
const int MAXL = 1000;
const int MAXC = 100;
char input[MAXN][MAXL+1];
char codes[10][MAXC+1];
int codes_len[10];

int matching_idx(char* input) {
    for (int i = 0; i < 10; i++)
        if (strncmp(input, codes[i], codes_len[i]) == 0)
            return i;
    // No matching found, should never happen
    assert(false);
}

int main() {
    int N;
    cin >> N;
    for (int i = 0; i < N; i++)
        cin >> input[i];
    for (int i = 0; i < 10; i++) {
        cin >> codes[i];
        codes_len[i] = strlen(codes[i]);
    }

    for (int i = 0; i < N; i++) {
        int start = 0, end = strlen(input[i]);
        while (start < end) {
            int idx = matching_idx(&input[i][start]);
            cout << idx;
            start += codes_len[idx];
        }
        // Consumed the whole input string
        assert(start == end);
        cout << endl;
    }
}

File: ois/ois20/round2/ois11_round2_solutions/ransomware.py
#!/usr/bin/env python3

# Starting from position 0, determines the (unique) match and advances
# the current position by the length of the matched codeword.

def matching_idx(text, codes):
    for idx, code in enumerate(codes):
        if text.startswith(code):
            return idx
    # No matching found, should never happen
    assert(False)

N = int(input().strip())
contacts = []
codes = []
for _ in range(N):
    contacts.append(input().strip())
for _ in range(10):
    codes.append(input().strip())
for contact in contacts:
    start = 0
    end = len(contact)
    while start < end:
        idx = matching_idx(contact[start:], codes)
        print(idx, end="")
        start += len(codes[idx])
    assert(start == end)
    print()
